Drivers/SensorFusion/Src/SensorFusion.cpp

```diff
@@ -0,0 +1,604 @@
+    mul(k, zMinHMultX, kMultZMinHMultX, DIM, NUM_MEASUREMENTS, 1);
+
+    float newX[DIM*1];
+    for (int i = 0; i < DIM*1; i++) newX[i] = x[i] + kMultZMinHMultX[i];
+
+    //Update error covariance: newP = (I - k*h)*p
+
+    //The identity matrix
+    float I[DIM*DIM] =
+    {
+        1, 0, 0, 0, 0, 0,
+        0, 1, 0, 0, 0, 0,
+        0, 0, 1, 0, 0, 0,
+        0, 0, 0, 1, 0, 0,
+        0, 0, 0, 0, 1, 0,
+        0, 0, 0, 0, 0, 1
+    };
+
+    float kMultH[DIM*DIM];
+    mul(k, h, kMultH, DIM, NUM_MEASUREMENTS, DIM);
+
+    float IMinKMultH[DIM*DIM];
+    for (int i = 0; i < DIM*DIM; i++) IMinKMultH[i] = I[i] - kMultH[i];
+
+    float newP[DIM*DIM];
+    mul(IMinKMultH, p, newP, DIM, DIM, DIM);
+
+
+    /*Output*/
+
+    double * latLongOut = cartesianToGPS(x[2], x[4]);
+
+    Output->altitude = x[0];
+    Output->rateOfClimb = x[1];
+    Output->latitude = latLongOut[0];
+    Output->latitudeSpeed = x[3];
+    Output->longitude = latLongOut[1];
+    Output->longitudeSpeed = x[5];
+
+    for (int i = 0; i < DIM*1; i++) iterdata->prevX[i] = newX[i];
+    for (int i = 0; i < DIM*DIM; i++) iterdata->prevP[i] = newP[i];
+
+    return SFError;
+}
+#endif
+
+
+SFError_t SF_GenerateNewResult()
+{
+    SFError_t SFError;
+    SFError.errorCode = 0;
+#ifdef   SF_Milestone3
+
+    SFAttitudeOutput_t attitudeOutput;
+    IMUData_t imuData;
+    imuObj->GetResult(imuData);
+
+    #ifdef  SF_Milestone4
+        SFPathOutput_t pathOutput;
+        GpsData_t GpsData;
+        AltimeterData_t altimeterData;
+        airspeedData_t airspeedData;
+        gpsObj->GetResult(GpsData);
+        altimeterObj->GetResult(altimeterData);
+        airspeedObj->GetResult(airspeedData);
+    
+
+
+        //Send gps timestamp
+        #ifndef UNIT_TESTING
+        TimeStampingPOGI *timeStamper = TimeStampingPOGI::GetInstance();
+
+        if(GpsData.ggaDataIsNew)
+        {
+            timeStamper->setGPSTime(&GpsData);
+        }
+        #endif
+    #endif
+
+    SFError = SF_GetAttitude(&attitudeOutput, &imuData);
+    if(SFError.errorCode != 0) return SFError;
+
+    SFOutput.pitch = attitudeOutput.pitch;
+    SFOutput.roll = attitudeOutput.roll;
+    SFOutput.yaw = attitudeOutput.yaw;
+
+    #ifdef  SF_Milestone4
+    SFError = SF_GetPosition(&pathOutput, &altimeterData, &GpsData, &imuData, &attitudeOutput, &iterData);
+    if(SFError.errorCode != 0) return SFError;
+
+    SFOutput.altitude = pathOutput.altitude;
+    SFOutput.rateOfClimb = pathOutput.rateOfClimb;
+    SFOutput.latitude = pathOutput.latitude;
+    SFOutput.latitudeSpeed = pathOutput.latitudeSpeed;
+    SFOutput.longitude = pathOutput.longitude;
+    SFOutput.longitudeSpeed = pathOutput.longitudeSpeed;
+    #endif
+
+#endif
+    return SFError;
+}
+
+
+
+SFError_t SF_GetResult(SFOutput_t *output)
+{
+    SFError_t SFError;
+
+    //TODO: This function always return success as return, fix it?
+    SFError.errorCode = 0;
+
+    *output = SFOutput;
+
+    return SFError;
```

Comment:
This does not seem thread safe. It simply returns a static pointer to the data that we put the updated data in.